dependent_types.md

```diff
@@ -0,0 +1,279 @@
+
+```
+data Void 
+data Unit = U
+data Bool = True | False
+```
+
+In contrast to this types can also have a _type level data_. That is the data that lives on the type level and does not have associated set of inhabitants. 
+
+Haskell is famous for its _If it compiles - it works_ approach, which is very true but we are haskellers - we always want more type safety and abstraction right ? 
+
+We can use type level data to provide extra safety to our programs since that allows us to _describe_ how our types should be constructed thus eliminating more runtime errors. 
+
+Here is a small example from `GHC` [manual](https://ghc.readthedocs.io) that provides a way of defining sort of an accessor with convenient type level _String_ (`Symbol`) 
+```
+{-# LANGUAGE DataKinds               #-}
+{-# LANGUAGE KindSignatures          #-}
+{-# LANGUAGE MultiParamTypeClasses   #-}
+{-# LANGUAGE FunctionalDependencies  #-}
+{-# LANGUAGE FlexibleInstances       #-}
+module Main where
+
+import GHC.TypeLits
+
+data Label (l :: Symbol) = Get
+data Point = Point Int Int deriving Show
+
+class Has a l b | a l -> b where
+  from :: a -> Label l -> b
+
+instance Has Point "x" Int
+    where from (Point x _) _ = x
+
+instance Has Point "y" Int
+    where from (Point _ y) _ = y
+
+example :: Int
+example = from (Point 1 2) (Get :: Label "x")
+
+```
+So type level data is something that lives on the level of types but is not a type.
+
+## Lambda Cube
+
+[Lambda Cube](https://en.wikipedia.org/wiki/Lambda_cube) is the term from Type Theory and Math Logic. It builds on top of [lambda calculus](https://en.wikipedia.org/wiki/Lambda_calculus) defining all possible type systems where each axes is the new abstraction. So going in some order from simple to more complex systems we first arrive to what every programming language has :
+
+**Values depending on values** (_functions_)
+- This simply means that you have some sort of relation between variables if they are used in terms of each other. 
+
+Example:
+
+```
+a :: Integer
+b = even a :: Bool
+-- here b depends on a 
+```
+
+After that we get to the second step which is
+
+**Values depending on Types** (_classes_)
+
+Example:
+
+```
+max :: Integer
+max :: Double
+```
+
+We can look at this as if type is determining the values which constant `max` can hold. This is a basic relation between values and types, values need to be of some type in Haskell.
+
+Third step in our wonderfull type land is 
+
+**Types depending on Types** (_type functions_)
+
+In order to demonstrate what we mean by this we will use Type Families. These [slides](https://cdepillabout.github.io/haskell-type-families-presentation/#/) do really good job in explaining what are type families.
+
+Basically they provide a way to define a function that operates on types. Depending on the concrete implementation we can return different types for different instances of the typeclass or at least we have that choice.
```

Comment:
Type families are completely separate from type classes (unless we're talking about associated types).